### PR TITLE
Use OrderedDictionary or Dictionary instead of SortedDictionary in RuntimeModel

### LINKED ISSUE #19361: ForeignKeys, Keys and Indexes on Property should have a strong order defined

### LINKED ISSUE #31142: Improve RuntimeModel building performance
Use `OrderedDictionary` as collection storage instead of `SortedDictionary`

### PR DESCRIPTION
**Pull request by AndriySvyryd**
Fixes #31142
Fixes #19361

Implementation copied from https://github.com/dotnet/corefxlab/blob/archive/src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs until it's included in the framework.

For reference, these are its perf characteristics:

Operation | Dictionary<K,V> | SortedDictionary<K,V> | SortedList<K,V> | OrderedDictionary<K,V>
-- | -- | -- | -- | --
this[key] | O(1) | O(log n) | O(log n) or O(n) | O(1)
this[index] | n/a | n/a | n/a | O(1)
Add(key,value) | O(1) | O(log n) | O(n) | O(1)
Remove(key) | O(1) | O(log n) | O(n) | O(n)
RemoveAt(index) | n/a | n/a | n/a | O(n)
ContainsKey(key) | O(1) | O(log n) | O(log n) | O(1)
ContainsValue(value) | O(n) | O(n) | O(n) | O(n)
IndexOf(key) | n/a | n/a | n/a | O(1)
Insert(index,key,value) | n/a | n/a | n/a | O(n)
Space efficiency | good | worst | medium | good
Insert/Remove sorted data | n/a | O(log n) | O(n) | n/a
From sorted data | n/a | ? | better? | n/a

### CODE REVIEW

**Comment by roji** [comment]
> Implementation copied from https://github.com/dotnet/corefxlab/blob/archive/src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/OrderedDictionary.cs until it's included in the framework.

Isn't it already [in the framework](https://learn.microsoft.com/en-us/dotnet/api/system.collections.specialized.ordereddictionary?view=net-7.0), or am I missing something?

**Comment by AndriySvyryd** [comment]
> Isn't it already [in the framework](https://learn.microsoft.com/en-us/dotnet/api/system.collections.specialized.ordereddictionary?view=net-7.0), or am I missing something?

Only the non-generic version. https://github.com/dotnet/runtime/issues/24826

**Comment by roji** [comment]
Oh interesting...

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs [hunk 1]
```diff
@@ -44,43 +44,39 @@ public override bool IsReadOnly
...
 
     /// <summary>
     ///     This is an internal API that supports the Entity Framework Core infrastructure and not subject to
     ///     the same compatibility standards as public APIs. It may be changed or removed without notice in
     ///     any release. You should only use it directly in your code with extreme caution and knowing that
     ///     doing so can result in application failures when updating to a new Entity Framework Core release.
     /// </summary>
-    public virtual SortedDictionary<(string, string?), Table> Tables { get; }
-        = new();
+    public virtual Dictionary<(string, string?), Table> Tables { get; } = new();
```
Comment (ajcvickers) [inline src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:55]: Should I run code cleanup again? ;-)

**Review by ajcvickers — APPROVED** [review]

File: src/EFCore.Relational/Metadata/Internal/RelationalModel.cs (hunk 1 above)
Comment (AndriySvyryd) [reply src/EFCore.Relational/Metadata/Internal/RelationalModel.cs:55]: You'll need to run it many times for 9.

**Pull request merged**